src/group_theory/is_free_group.lean

```diff
@@ -0,0 +1,94 @@
+/-
+Copyright (c) 2021 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+import group_theory.free_group
+/-!
+This file defines the universal property of free groups, and proves some things about
+groups with this property. For an explicit construction of free groups, see
+`group_theory/free_group`.
+-/
+noncomputable theory
+universes u w
+
+/-- `is_free_group G` means that `G` has the universal property of a free group.
+    That is, it has a family `generators G` of elements, such that a group homomorphism
+    `G →* X` is uniquely determined by a function `generators G → X`. -/
+class is_free_group (G) [group.{u} G] : Type (u+1) :=
+(generators : Type u)
+(of : generators → G)
+(unique_lift : ∀ {X} [group.{u} X] (f : generators → X),
+                ∃! F : G →* X, ∀ a, F (of a) = f a)
+
+instance free_group_is_free_group {A} : is_free_group (free_group A) :=
+{ generators := A,
+  of := free_group.of,
+  unique_lift := begin
+    introsI X _ f,
+    exact ⟨free_group.lift f,
+      λ _, free_group.lift.of,
+      λ g hg, monoid_hom.ext (λ _, free_group.lift.unique g hg)⟩
+  end }
+
+namespace is_free_group
+
+variables {G H : Type u} [group G] [group H] [is_free_group G]
+
+/-- The group homomorphism from a free group given by some function on the generators. -/
+def lift' (f : generators G → H) : G →* H := classical.some (unique_lift f)
```

Comment:
Is there an advantage to going noncomputable here rather than having `lift : (generators A → H) ≃ (A →* H)` be a field of `is_free_group`?

Reply:
There's an argument we should go the other direction --- aiming to make `is_free_group` live in `Prop` (so typeclass inference couldn't ever find incompatible instances), by defining it as something like
"there exists a `S : set G` so that the map restricting a `G →* X` to `S → X` is a bijection".

Reply:
And then define `generators G` via choice.

Reply:
> typeclass inference couldn't ever find incompatible instances

Presumably the case you're thinking of is one where the group is free over two different sets of generators? One way to resolve that would be to change the typeclass to:
```
class is_free_group (G) (generators : out_param (Type u)) [group G]
```

The advantage of having it live in `Type` is it allows things like "my object is a free group so define a computable function by starting from its generators".

Reply:
I suppose that might be better suited to a name like `has_free_group_lift` or something though

Reply:
Regarding your first question, I don't think it matters much whether the lift is computable or not. I mean, even if it's marked as noncomputable, `simp` can still compute with it since we know how it behaves on generators. On the other hand it should be easy to make this field data if you prefer.

Scott's suggestion seems a bit extreme in this case. There's a difference between "the group Z of integers is freely generated by 1" and "Z is free", and I expect the former statement is usually what you want. (The Nielsen-Schreier index theorem would also be tricky to prove if `generators` were given by choice.) I don't expect we'll see to many instances of `is_free_group`, so hopefully incompatible instances won't be too big a problem.

Unbundling generators (and the inclusion `of : A -> G` of your generators to your group) would work, but it's a bit awkward having to write `is_free_group G A of` instead of `is_free_group G` (especially in the statement of Nielsen-Schreier, where you'd have to move the generating set of your subgroup from the proof of the theorem to the statement).

Reply:
I'll say a bit more about different generating sets: with Z you can of course show that any generating set is {1} or {-1}. But with the free group on two generators a, b you have another generating set {a, ab}, and also {aba, ab}, etc. There's no way to control which one is picked by `classical.choice`.

Reply:
Great, I agree with your analysis, and I'm happy leaving the definition as you've done it.